src/client/cursor.rs

```diff
@@ -0,0 +1,171 @@
+use bson;
+use client::wire_protocol::flags::OpQueryFlags;
+use client::wire_protocol::operations::Message;
+use std::collections::vec_deque::VecDeque;
+use std::io::{Read, Write};
+
+/// Maintains a connection to the server and lazily returns documents from a
+/// query.
+///
+/// # Fields
+///
+/// `request_id` - Uniquely identifies the request being sent.
+/// `namespace` - The full qualified name of the collection,
+///                          beginning with the database name and a period.
+/// `batch_size` - How many documents to fetch at a given time from the server.
+/// `cursor_id` - Uniquely identifies the cursor being returned by the reply.
+pub struct Cursor<'a, T> where T: Read + Write + 'a {
+    request_id: i32,
+    namespace: String,
+    batch_size: i32,
+    cursor_id: i64,
+    limit: i32,
+    count: i32,
+    buffer: VecDeque<bson::Document>,
+    stream: &'a mut T,
+}
+
+impl <'a, T> Cursor<'a, T> where T: Read + Write + 'a {
+    /// Gets the cursor id and BSON documents from a reply Message.
+    ///
+    /// # Arguments
+    ///
+    /// `message` - The reply message to get the documents from.
+    ///
+    /// # Return value.
+    ///
+    ///
+    fn get_bson_and_cid_from_message(message: Message) -> Option<(VecDeque<bson::Document>, i64)> {
+        match message {
+            Message::OpReply { header: _, flags: _, cursor_id: cid,
+                               starting_from: _, number_returned: _,
+                               documents: docs } => {
+                let mut v = VecDeque::new();
+
+                for bson_doc in docs {
```

Comment:
nitpick: `doc` is probably better than `bson_doc`, slightly redundant to name something `bson_doc` when that's the only type of document we're supporting for the foreseeable future.